### PR TITLE
type `ArraySlice` support spread array #828

### LINKED ISSUE #828: `ArraySlice` should support spread array
v4.12.0
```ts
const a: ArraySlice<[1, 2, 3, ...string[]], 1>
// => []
```

expect:
```ts
const a: ArraySlice<[1, 2, 3, ...string[]], 1>
// => [2, 3, ...string[]]
```

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/828">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/828/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Max10240**
Fixes #828

### CODE REVIEW

**Comment by Max10240** [comment]
@Emiyaaaaa review this pls :)

**Comment by sindresorhus** [comment]
@Max10240 Tip for the future, make pull requests from a feature branch in your fork, not the `main` branch.

**Comment by Max10240** [comment]
> @Max10240 Tip for the future, make pull requests from a feature branch in your fork, not the `main` branch.

fine

File: test-d/array-slice.ts [hunk 1]
```diff
@@ -18,3 +18,11 @@ expectType<ArraySlice<[0, 1, 2, 3], 0, -3>>([0]);
...
 expectType<ArraySlice<[], -10, 1>>([]);
 expectType<ArraySlice<[]>>([]);
+
+expectType<ArraySlice<[1, 2, 3, ...string[]], 0, 3>>([1, 2, 3]);
+expectType<ArraySlice<[1, 2, 3, ...string[]], 1, 5>>([2, 3, null! as string, null! as string]);
+expectType<ArraySlice<[1, 2, 3, ...string[], 4, 5], 1, 5>>([2, 3, null! as (string | 4 | 5), null! as (string | 4 | 5)]);
+expectType<ArraySlice<[1, 2, 3, ...string[], 4, 5], 0>>([1, 2, 3, ...(null! as string[]), 4, 5]);
+expectType<ArraySlice<[1, 2, 3, ...string[], 4, 5], 1>>([2, 3, ...(null! as string[]), 4, 5]);
+expectType<ArraySlice<[1, 2, 3, ...string[], 4, 5], 3>>([...(null! as string[]), 4, 5]);
+expectType<ArraySlice<[1, 2, 3, ...string[], 4, 5], 10>>([...(null! as string[]), 4, 5]);
```
Comment (Emiyaaaaa) [inline test-d/array-slice.ts:28]: could you add a test for leading spread array?

like `expectType<ArraySlice<[...string[], 1, 2, 3], 1, 3>`

Comment (Max10240) [reply test-d/array-slice.ts:28]: sure！

**Pull request closed without merging**